### PR TITLE
NO-149/SurfaceRequester

### PR DESCRIPTION
**Pull request by Mecharyry**
## Problem
The current implementation of `NoPlayer` relies upon a `SurfaceHolderRequester` to pass a `SurfaceHolder` to the underlying player. Clients may want to use `SurfaceView`, `TextureView` etc and the underlying player, `ExoPlayer` specifically only cares about the `Surface`.

## Solution
Switch out the `SurfaceHolderRequester` with a `SurfaceRequester` passing a `Surface` to the underlying player. 

### Paired with 
Nobody.

### CODE REVIEW

File: core/src/main/java/com/novoda/noplayer/FramesPerSecondCalculator.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package com.novoda.noplayer;
+
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.concurrent.TimeUnit;
+
```
Comment (ouchadam) [inline core/src/main/java/com/novoda/noplayer/FramesPerSecondCalculator.java]: should this class be in this PR?

Comment (mr-archano) [reply core/src/main/java/com/novoda/noplayer/FramesPerSecondCalculator.java]: Yeah, I would consider to remove it @Mecharyry, just to avoid to ship it in what is effectively a non-related change set.

File: core/src/main/java/com/novoda/noplayer/SurfaceRequester.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package com.novoda.noplayer;
+
+import android.view.Surface;
+
+public interface SurfaceRequester {
+
+    void requestSurfaceTexture(Callback callback);
```
Comment (mr-archano) [inline core/src/main/java/com/novoda/noplayer/SurfaceRequester.java]: No need to use texture, right?

File: core/src/main/res/layout/noplayer_view.xml [hunk 3]
```diff
@@ -5,7 +5,7 @@
   android:layout_height="wrap_content">
 
   <SurfaceView
-    android:id="@+id/surface_view"
+    android:id="@+id/texture_view"
```
Comment (mr-archano) [inline core/src/main/res/layout/noplayer_view.xml]: shouldn't we keep this as it was?

File: demo/src/main/java/com/novoda/demo/MainActivity.java [hunk 4]
```diff
@@ -75,17 +75,20 @@ public void onClick(View v) {
     };
 
     private final View.OnClickListener showAudioSelectionDialog = new View.OnClickListener() {
-
         @Override
         public void onClick(View v) {
-            dialogCreator.showAudioSelectionDialog();
+            if (player == null || !player.isPlaying() || player.getAudioTracks().size() == 0) {
+                Toast.makeText(MainActivity.this, "no audio tracks available!", Toast.LENGTH_LONG).show();
```
Comment (mr-archano) [inline demo/src/main/java/com/novoda/demo/MainActivity.java]: Why this change?

File: demo/src/main/java/com/novoda/demo/MainActivity.java [hunk 5]
```diff
@@ -75,17 +75,20 @@ public void onClick(View v) {
...
+                dialogCreator.showAudioSelectionDialog();
+            }
         }
     };
 
     private final View.OnClickListener showSubtitleSelectionDialog = new View.OnClickListener() {
         @Override
         public void onClick(View v) {
-            if (player.getSubtitleTracks().isEmpty()) {
+            if (player == null || !player.isPlaying() || player.getSubtitleTracks().isEmpty()) {
```
Comment (mr-archano) [inline demo/src/main/java/com/novoda/demo/MainActivity.java]: Same question here (I know that this is a demo, I am just curious)

File: demo/src/main/java/com/novoda/demo/MainActivity.java (hunk 4 above)
Comment (Mecharyry) [reply demo/src/main/java/com/novoda/demo/MainActivity.java]: The buttons to trigger the dialog are always present, if you tap them before the player is playing then it will throw an `IllegalStateException`.

File: demo/src/main/java/com/novoda/demo/MainActivity.java (hunk 5 above)
Comment (Mecharyry) [reply demo/src/main/java/com/novoda/demo/MainActivity.java]: Same as above 😄

File: core/src/main/java/com/novoda/noplayer/FramesPerSecondCalculator.java (hunk 1 above)
Comment (Mecharyry) [reply core/src/main/java/com/novoda/noplayer/FramesPerSecondCalculator.java]: 728d499

File: core/src/main/java/com/novoda/noplayer/SurfaceRequester.java (hunk 2 above)
Comment (Mecharyry) [reply core/src/main/java/com/novoda/noplayer/SurfaceRequester.java]: 20b70e5

File: core/src/main/res/layout/noplayer_view.xml (hunk 3 above)
Comment (Mecharyry) [reply core/src/main/res/layout/noplayer_view.xml]: 899678b

File: core/src/main/java/com/novoda/noplayer/NoPlayerView.java [hunk 6]
```diff
@@ -51,7 +51,7 @@ public View getContainerView() {
     }
 
     @Override
-    public SurfaceHolderRequester getSurfaceHolderRequester() {
+    public SurfaceRequester getSurfaceTextureRequester() {
```
Comment (ouchadam) [inline core/src/main/java/com/novoda/noplayer/NoPlayerView.java]: Do we need to leak `texture` here? SurfaceRequester should be enough?

Comment (Mecharyry) [reply core/src/main/java/com/novoda/noplayer/NoPlayerView.java]: 6312a81
